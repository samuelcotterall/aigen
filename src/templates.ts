import { Eta } from "eta";
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { AgentConfig } from "./schema.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TEMPLATES_DIR = path.join(__dirname, "../templates");

type FrontMatter = {
  when?: Partial<{ preset: string; libraries: string[] }>;
  emit?: boolean;
  outPath?: string;
};

const FRONT_RE = /^---json\s*\n([\s\S]*?)\n---\s*\n?/;

/**
 * Parse a JSON-front-matter block from a template file.
 *
 * Templates may start with a block like:
 * ---json
 * { "when": { "preset": "openai" }, "outPath": "agent.md" }
 * ---
 *
 * @param src - raw template source
 * @returns parsed front matter and the remaining body
 */
function parseFrontMatter(src: string): { fm: FrontMatter; body: string } {
  const m = src.match(FRONT_RE);
  if (!m) return { fm: {}, body: src };
  const fm = JSON.parse(m[1] || "{}") as FrontMatter;
  return { fm, body: src.slice(m[0].length) };
}

/**
 * Test whether a template's `when` conditions match the provided config.
 *
 * @param cfg - the current AgentConfig
 * @param when - optional front-matter `when` constraints
 */
function matchesWhen(cfg: AgentConfig, when?: FrontMatter["when"]) {
  if (!when) return true;
  if (when.preset && when.preset !== (cfg.preset ?? "")) return false;
  if (when.libraries && when.libraries.length) {
    const set = new Set(cfg.libraries ?? []);
    if (!when.libraries.some((x) => set.has(x as any))) return false;
  }
  return true;
}

async function walk(dir: string): Promise<string[]> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  const files: string[] = [];
  for (const e of entries) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) files.push(...(await walk(p)));
    else files.push(p);
  }
  return files;
}

/**
 * Render all templates to a map of output path -> content.
 *
 * This will evaluate Eta templates under `templates/common` and return a
 * flat object containing the final content for each output path.
 *
 * @param cfg - parsed AgentConfig used as the template context
 */
export async function renderTemplates(
  cfg: AgentConfig,
  opts?: { emitAgents?: boolean }
) {
  // Enable `useWith: true` so provider templates that reference `cfg` at the
  // top-level continue to work. When rendering we also pass an `it` object
  // containing `cfg` so templates that use `it.cfg` (the common templates)
  // still function correctly.
  const eta = new Eta({ views: TEMPLATES_DIR, useWith: true });
  // Provide sane defaults so calling code (tests, scripts) can pass a small
  // cfg object without needing to populate every nested field that templates
  // might reference.
  const defaultStyle = { tsconfig: "strict", naming: "kebab", docs: "typedoc" };
  const c: any = cfg as any;
  const renderCfg: AgentConfig = {
    name: c.name,
    displayName: c.displayName,
    preset: c.preset,
    libraries: c.libraries ?? [],
    tools: c.tools ?? [],
    policies: c.policies ?? {},
    style: { ...(c.style ?? {}), ...defaultStyle },
    styleRules: c.styleRules ?? [],
  } as any;
  const all = await walk(TEMPLATES_DIR);
  const outputs: Record<string, string> = {};

  for (const abs of all) {
    const rel = path.relative(TEMPLATES_DIR, abs);
    const raw = await fs.readFile(abs, "utf8");
    const { fm, body } = parseFrontMatter(raw);
    // Handle provider-specific templates under templates/agents/<provider>/...
    const parts = rel.split(path.sep);
    if (rel.startsWith("agents/")) {
      const provider = parts[1] || undefined;
      const hasWhen = !!fm.when;
      const candidate = fm.emit === true || (opts && opts.emitAgents === true);
      // By default skip provider-specific templates unless explicitly requested
      // via opts.emitAgents or per-template front-matter `emit: true`.
      if (!candidate) continue;

      // Determine whether there are `when` constraints to enforce. If the
      // template provided explicit `when`, enforce it. If we are emitting due
      // to the `emitAgents` option, infer the preset from the provider folder
      // and enforce that. If the template had `emit: true` but no `when`, do
      // not enforce a preset match.
      let effectiveWhen = fm.when;
      if (!effectiveWhen && opts && opts.emitAgents && provider) {
        effectiveWhen = { preset: provider };
      }
      if (effectiveWhen && !matchesWhen(renderCfg, effectiveWhen)) continue;
    } else {
      if (!matchesWhen(renderCfg, fm.when)) continue;
    }

    // Place files from `templates/common/**` at the root of the generated pack
    const defaultOut = rel.replace(/^common\//, "").replace(/\.eta$/, "");
    const outPath = fm.outPath ?? defaultOut;
    // Provide both `cfg` (for provider templates) and `it: { cfg }` so both
    // styles of templates work: those that directly reference `cfg` and those
    // that expect `it.cfg`.
    const rendered = eta.renderString(body, {
      cfg: renderCfg,
      it: { cfg: renderCfg } as any,
    });
    if (typeof rendered !== "string") continue;
    outputs[outPath] = rendered;
  }
  return outputs;
}
