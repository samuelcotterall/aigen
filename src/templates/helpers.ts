import crypto from "node:crypto";

// Small helpers for templates that rely on a provided RNG function.
// `rng` must be a zero-arg function returning a float in [0,1).
export function makeTemplateHelpers(rng: () => number) {
  let counter = 0;
  return {
    // deterministic short id: prefix + base36(counter + random)
    idFromRng(prefix = "id") {
      const n = Math.floor(rng() * 0xffffffff) + counter++;
      return `${prefix}-${n.toString(36)}`;
    },
    // deterministic pseudo-uuid from rng by hashing random bytes
    uuidFromRng() {
      const parts: string[] = [];
      for (let i = 0; i < 4; i++) {
        const n = Math.floor(rng() * 0xffffffff);
        parts.push(n.toString(16).padStart(8, "0"));
      }
      const seed = parts.join("");
      // create a short, stable hex digest
      return crypto.createHash("sha1").update(seed).digest("hex").slice(0, 32);
    },
    // convenience: produce a hex string from rng
    hex(nBytes = 8) {
      const parts: string[] = [];
      for (let i = 0; i < nBytes; i++) {
        parts.push(Math.floor(rng() * 256).toString(16).padStart(2, "0"));
      }
      return parts.join("");
    },
  } as const;
}
