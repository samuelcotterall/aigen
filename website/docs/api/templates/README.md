[**create-agent-instructions**](/docs/api)

***

[create-agent-instructions](/docs/api/modules) / templates

# templates

## Functions

- [renderTemplates](functions/renderTemplates.md)
