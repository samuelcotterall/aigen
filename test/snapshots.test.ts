import { describe, it, expect } from "vitest";
import { renderTemplates } from "../src/templates.js";

describe("render snapshots", () => {
  it("agent and docs match snapshot for sample config", async () => {
    const cfg: any = {
      name: "snapshot-agent",
      displayName: "Snapshot Agent",
      preset: "openai",
      libraries: ["react"],
      tools: [{ name: "fetch", description: "HTTP" }],
      policies: { safetyLevel: "low" },
      style: { tsconfig: "strict", naming: "kebab", docs: "typedoc" },
      styleRules: [
        {
          id: "r1",
          label: "[sec] r1",
          description: "desc r1",
          doText: "do r1",
          dontText: "dont r1",
          appliesTo: ["react"],
        },
      ],
    };

    const out = await renderTemplates(cfg as any);

    // snapshot the main outputs (per-template snapshots)
    expect(out["agent.md"]).toMatchSnapshot("agent.md");
    expect(out["tools.md"]).toMatchSnapshot("tools.md");
    expect(out["policies.md"]).toMatchSnapshot("policies.md");
    expect(out["EXAMPLES.md"]).toMatchSnapshot("EXAMPLES.md");
  });
});
