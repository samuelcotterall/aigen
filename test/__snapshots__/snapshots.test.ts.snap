// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render snapshots > agent and docs match snapshot for sample config > EXAMPLES.md 1`] = `
"# Examples for
This file demonstrates sample usage patterns for different presets.

## OpenAI examples
- See \`examples/openai.ts\` for a minimal runtime example.

"
`;

exports[`render snapshots > agent and docs match snapshot for sample config > agent.md 1`] = `
"
# Snapshot Agent — System Instructions

- Runtime: openai
- Conventions: strict TS, naming: kebab, docs: typedoc
You are Snapshot Agent, an AI agent that plans and executes tasks via tools.

## Operating principles
1) Be explicit about assumptions
2) Prefer small, reversible steps
3) Summarize tool results briefly with sources if applicable

- **fetch** — HTTP
## Style rules

The following style rules are enabled for this agent. Each rule includes a short description and practical Do / Don't guidance.

### [sec] r1_Applies to:_
- react
**Description:** desc r1
**Do:**
- do r1
**Don't:**
- dont r1
"
`;

exports[`render snapshots > agent and docs match snapshot for sample config > policies.md 1`] = `
"# # Snapshot Agent — Policies
- Safety: low- Refusals: undefined- Privacy: Avoid unnecessary data retention.
## Style rules

Enabled rules:
- [sec] r1"
`;

exports[`render snapshots > agent and docs match snapshot for sample config > tools.md 1`] = `
"# Snapshot Agent — Tools
Each tool has a JSON input and optional JSON output.

## fetch
**Description:** HTTP
**Input:**
\`\`\`json
{}
\`\`\`

## Style rules

Enabled rules:
- [sec] r1"
`;
